guestlink.php

```diff
@@ -0,0 +1,93 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Page to grant external users access to a BBB session
+ *
+ * @package    mod_bigbluebuttonbn
+ * @author     Angela Baier
+ * @copyright  2020 University of Vienna
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+require('../../config.php');
+require_once(__DIR__ . '/lib.php');
+require_once(__DIR__ . '/locallib.php');
+
+global $PAGE, $OUTPUT;
+
+$gid = required_param('gid', PARAM_ALPHANUM); // This is required.
+$guestname = optional_param('guestname', '', PARAM_TEXT);
+$guestpass = optional_param('guestpass', '', PARAM_TEXT);
+$PAGE->set_url(new moodle_url('/mod/bigbluebuttonbn/guestlink.php',
+        ['gid' => $gid, 'guestname' => $guestname, 'guestpass' => $guestpass]));
+$PAGE->set_context(context_system::instance());
+$PAGE->set_pagelayout('embedded');
+
+if (!\mod_bigbluebuttonbn\locallib\config::get('participant_guestlink')) {
+    echo get_string('guestlink_form_guestlink_disabled', 'bigbluebuttonbn');
+    die();
+}
+
+$bigbluebuttonbn = bigbluebuttonbn_get_bigbluebuttonbn_by_guestlinkid($gid);
+if (!$bigbluebuttonbn->guestlinkenabled) {
+    echo get_string('guestlink_form_guestlink_disabled_instance', 'bigbluebuttonbn');
+    die();
+}
+$valid = ($guestname && ($bigbluebuttonbn->guestpass == $guestpass || !$bigbluebuttonbn->guestpass));
+
+if (!$valid) {
+    $guestpasserrormessage = false;
+    $guestnameerrormessage = false;
+    if ($guestpass && $bigbluebuttonbn->guestpass != $guestpass) {
+        $guestpasserrormessage = true;
+    }
+    $customlogo = \mod_bigbluebuttonbn\locallib\config::get('custom_logo');
```

Comment:
I am not found on this `custom_logo` idea. I understand that the idea is to differenciate the guestlink page from the rest to let the user know that this is for entering BBB, but most Universities/Colleges actually customize or ask customization to their theme. Why would we force them to do different if they have already their theme?

I think the guestlink page should be as any other moodle page. Perhaps like the Sign in page.

Reply:
We had a discussion about this ourselves. There are some things to consider:
a) it is not possible to use the standard moodle page since guests don't have access to any other content in moodle. So they can't see the navigation and so on.
b) not everyone has a signin page, since many sites are behind some SSO. Since this is the case we can't rely on any login page to be working. Even if there is a login page: The theme can change the way it is processed and there is no real way to ensure that a hook or something like that at least works on every moodle instance.
So we oriented on greenlight. And in greenlight you at least have the chance to change the logo, which is something I have seen on some other moodle plugins as well. Anyone can still change the CSS of this page, it's one of the most simplistic ones I have seen in my carreer and should just take some minutes.
If you have an idea how to solve the issue I am open for suggestions.

Reply:
The intention of the 'embedded' layout is for use inside of iframes and similar. I don't understand why a normal page layout can't be used, in particular the 'login' layout, or if that isn't good then 'base' or 'secure' should work. There are plenty of other similar page which are visible to the public, even if moodle guest mode is switched off.